### PR TITLE
Enhancement to DNS/DHCP/SSH support for virtualization tests

### PR DESCRIPTION
**Pull request by waynechen55**
Enhance data/virt_autotest/setup_dns_service.sh by
1) Get domain information through parameters, improve help info, check invalid parameters, make some parameters mandatory
2) Sign modified config file by using script name, restore tainted config files if it was modified previously
3) Restart dhcpd service with updated configurations
4) Retry to ping/ssh-copy-id vm guest after specific interval instead of sleep a amoount of time in order to speed up script execution
5) More strict termination mechanism if any failure
6) Add more explanation comments to help understand

- Related ticket: n/a
- Needles: n/a
- Verification run: 
   Passed http://10.67.18.219/tests/279
   Failed http://10.67.18.219/tests/281

### CODE REVIEW

**Comment by waynechen55** [comment]
Please have a look if any concerns @alice-suse @guoxuguang

File: data/virt_autotest/setup_dns_service.sh [hunk 1]
```diff
@@ -174,48 +254,165 @@ for single_nameserver in ${nameservers_array[@]};do
...
 
-#Ping vm machines one by one. Quit if any failure.
-sleep 45s
+#Generate ssh key and set it as IdentityFile in /etc/ssh/ssh_config
+ssh_key_pass=`/usr/share/qa/virtautolib/lib/get-settings.sh vm.pass`
+ssh_key_path="/var/${dns_domain_forward}/.ssh"
+ssh_config_file="/etc/ssh/ssh_config"
+ssh_config_temp="/etc/ssh/ssh_config.temp"
+cat ${ssh_config_file} | grep "#$0"
+if [ `echo $?` == 0 ];then
```
Comment (alice-suse) [inline data/virt_autotest/setup_dns_service.sh]: you can directly use ```if [ $? -eq 0 ];then```

File: tests/virt_autotest/setup_dns_service.pm [hunk 2]
```diff
@@ -18,19 +18,25 @@ use testapi;
...
+    my $wait_script          = "180";
+    my $dns_bash_script_name = "setup_dns_service.sh";
+    my $dns_bash_script_url  = data_url("virt_autotest/$dns_bash_script_name");
+    my $dns_forward_domain   = "testvirt.net";
+    my $dns_reverse_domain   = "123.168.192";
+    my $dns_server_ipaddr    = "192.168.123.1";
+    my $download_bash_script = "curl -s -o ~/$dns_bash_script_name $dns_bash_script_url";
+    script_output($download_bash_script, $wait_script, type_command => 0, proceed_on_failure => 0);
+    my $execute_bash_script = "chmod +x ~/$dns_bash_script_name && ~/$dns_bash_script_name -f $dns_forward_domain -r $dns_reverse_domain -s $dns_server_ipaddr";
+    script_output($execute_bash_script, $wait_script, type_command => 0, proceed_on_failure => 0);
```
Comment (alice-suse) [inline tests/virt_autotest/setup_dns_service.pm]: if this execute_bash_script failed, will the post_fail_hook be triggered? I plan to check this in failure job, but seems the given link can not be opened. Would you please update the link http://10.67.18.219/tests/281?

File: data/virt_autotest/setup_dns_service.sh (hunk 1 above)
Comment (waynechen55) [reply data/virt_autotest/setup_dns_service.sh]: Done

File: tests/virt_autotest/setup_dns_service.pm (hunk 2 above)
Comment (waynechen55) [reply tests/virt_autotest/setup_dns_service.pm]: It is valid. You can also refer to http://147.2.217.156/tests/281  or http://147.2.217.156/tests/280   If none works for you, I think that is magic.

Comment (alice-suse) [reply tests/virt_autotest/setup_dns_service.pm]: I can open the two jobs you give. Yes, post_fail_hook is triggered.

**Comment by guoxuguang** [comment]
LGTM

**Pull request merged**